File: 2169-simple-bank-system/2169-simple-bank-system.cpp
#pragma GCC optimimze("O3", "unroll-loops")
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
class Bank {
   private:
    vector<long long> balance;
    int n;
    bool validate(int account) {
        if (account >= 1 && account <= n) {
            return true;
        }
        return false;
    }

   public:
    Bank(vector<long long>& balance) {
        this->balance = balance;
        this->n = balance.size();
    }

    bool transfer(int account1, int account2, long long money) {
        if (validate(account1) && validate(account2) &&
            balance[account1 - 1] >= money) {
            balance[account1 - 1] -= money;
            balance[account2 - 1] += money;
            return true;
        }
        return false;
    }

    bool deposit(int account, long long money) {
        if (validate(account)) {
            balance[account - 1] += money;
            return true;
        }
        return false;
    }

    bool withdraw(int account, long long money) {
        if (validate(account) && balance[account - 1] >= money) {
            balance[account - 1] -= money;
            return true;
        }
        return false;
    }
};

/**
 * Your Bank object will be instantiated and called as such:
 * Bank* obj = new Bank(balance);
 * bool param_1 = obj->transfer(account1,account2,money);
 * bool param_2 = obj->deposit(account,money);
 * bool param_3 = obj->withdraw(account,money);
 */
/**
 * Your Bank object will be instantiated and called as such:
 * Bank* obj = new Bank(balance);
 * bool param_1 = obj->transfer(account1,account2,money);
 * bool param_2 = obj->deposit(account,money);
 * bool param_3 = obj->withdraw(account,money);
 */

File: 2169-simple-bank-system/2169-simple-bank-system.go
type Bank struct {
    bal []int64 
    n int
}


func Constructor(balance []int64) Bank {
    return Bank{bal: balance, n: len(balance)} 
}

func (this *Bank) validate(acc int) bool {
    return acc >= 1 && acc <= this.n
}

func (this *Bank) Transfer(account1 int, account2 int, money int64) bool {
    if this.validate(account1)  && this.validate(account2) && this.bal[account1-1] >= money {
        this.bal[account1-1] -= money
        this.bal[account2-1] += money
        return true
    }
    return false
}


func (this *Bank) Deposit(account int, money int64) bool {
    if this.validate(account) {
        this.bal[account-1] += money
        return true
    } 
    return false
}


func (this *Bank) Withdraw(account int, money int64) bool {
    if this.validate(account) && this.bal[account-1] >= money {
        this.bal[account-1] -= money
        return true
    }
    return false
}


/**
 * Your Bank object will be instantiated and called as such:
 * obj := Constructor(balance);
 * param_1 := obj.Transfer(account1,account2,money);
 * param_2 := obj.Deposit(account,money);
 * param_3 := obj.Withdraw(account,money);
 */